syncer/data_fetch.go

```diff
@@ -67,41 +55,36 @@ func (d *DataFetch) PollLayerData(ctx context.Context, lid types.LayerID, peers
 	logger := d.logger.With(zap.Uint32("layer", lid.Uint32()), log.ZContext(ctx))
 	layerData := make(chan fetch.LayerData, len(peers))
 	var eg errgroup.Group
-	fetchErr := threadSafeErr{}
 	for _, peer := range peers {
 		eg.Go(func() error {
 			data, err := d.fetcher.GetLayerData(ctx, peer, lid)
 			if err != nil {
 				layerPeerError.Inc()
 				logger.With().Debug("failed to get layer data", zap.Error(err), zap.Stringer("peer", peer))
-				fetchErr.join(err)
-				return nil
+				return err
```

Comment:
Same here, the `ctx` passed to `GetLayerData` could be from `errgroup.WithContext` so that all fetches fail as soon as the first fails.

Reply:
We don't want to cancel the other fetches running in parallel. We fetch from many peers for redundancy and we need only one to succeed for overall success. Check the success condition below.